src/v/cloud_storage/remote.cc

```diff
@@ -715,6 +730,82 @@ ss::future<upload_result> remote::delete_objects(
   std::vector<cloud_storage_clients::object_key> keys,
   retry_chain_node& parent) {
     ss::gate::holder gh{_gate};
+
+    if (!is_batch_delete_supported()) {
+        vlog(
+          cst_log.info,
+          "Backend {} does not support batch delete, falling back to "
+          "sequential deletes",
+          _cloud_storage_backend);
+
+        auto keys_empty = keys.empty();
+
+        // Holds the key and associated retry_chain_node to make it
+        // easier to keep objects on heap using do_with, as retry_chain_node
+        // cannot be copied or moved.
+        struct key_and_node {
+            cloud_storage_clients::object_key key;
+            std::unique_ptr<retry_chain_node> node;
+        };
+
+        std::vector<key_and_node> key_nodes;
+        key_nodes.reserve(keys.size());
+
+        std::transform(
+          std::make_move_iterator(keys.begin()),
+          std::make_move_iterator(keys.end()),
+          std::back_inserter(key_nodes),
+          [&parent](auto&& key) {
+              return key_and_node{
+                .key = std::forward<cloud_storage_clients::object_key>(key),
+                .node = std::make_unique<retry_chain_node>(&parent)};
+          });
+
+        auto results = co_await ss::do_with(
+          bucket,
+          std::move(key_nodes),
+          std::vector<upload_result>{},
+          [this](auto& bucket, auto& key_nodes, auto& results) {
+              results.reserve(key_nodes.size());
+              return ss::max_concurrent_for_each(
+                       key_nodes.begin(),
+                       key_nodes.end(),
+                       concurrency(),
```

Comment:
Using the full `concurrency()` here will be quite aggressive if we use this method from partition deletion/GC there are multiple partitions undergoing deletion/GC at the same time: they'll all be trying to use the full number of connections that the shard has.

It's okay for this PR, but we may later need to pass in an advisory concurrency argument to the function, as something like a topic delete can figure out how many partitions are deleted in parallel on the same shard.